jelly/src/main/java/org/kohsuke/stapler/framework/adjunct/AdjunctManager.java

```diff
@@ -157,13 +158,13 @@ public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOExceptio
         String path = req.getRestOfPath();
         if (path.charAt(0)=='/') path = path.substring(1);
 
-        if(!allowedResources.containsKey(path)) {
+        if(!allowedResources.contains(path)) {
             if(!allowResourceToBeServed(path)) {
                 rsp.sendError(SC_FORBIDDEN);
                 return;
             }
             // remember URLs that we can serve. but don't remember error ones, as it might be unbounded
-            allowedResources.put(path,path);
+            allowedResources.add(path);
```

Comment:
No this does not fix the issue at all—it is still doing _two_ operations. That is why I suggested using the return value of `add`: it atomically adds the path if not already present, and lets you know if it did so or not.

Reply:
OK, needed to reread

> don't remember error ones, as it might be unbounded

IOW this is not a regular cache and cannot use the regular idioms.